### PR TITLE
Modify OCSP to use alt MD for cert IDs in responses

### PR DESCRIPTION
**Pull request by tmshort**
Modify openssl OCSP utility to produce certIDs in responses using other
hash algorithms (e.g. SHA256).

Added option -rcid for this purpose.

##### Checklist

- [x] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: apps/ocsp.c [hunk 1]
```diff
@@ -594,7 +603,8 @@ int ocsp_main(int argc, char **argv)
 
     if (rdb != NULL) {
         make_ocsp_response(bio_err, &resp, req, rdb, rca_cert, rsigner, rkey,
-                               rsign_md, rsign_sigopts, rother, rflags, nmin, ndays, badsig);
+                               rsign_md, rsign_sigopts, rother, rflags, nmin, ndays, badsig,
```
Comment (richsalz) [inline apps/ocsp.c]: as long as you're hear, make the indent line under the bio_err param.

**Comment by tmshort** [comment]
This might duplicate functionality added in commit 4dd21a7b8b850a39b0f610fceca21557853c943
so it might not be necessary..?

**Comment by tmshort** [comment]
Actual link to commit
https://github.com/openssl/openssl/commit/b4dd21a7b8b850a39b0f610fceca21557853c943

**Comment by tmshort** [comment]
According to our internal tester, this does not duplicate functionality of commit b4dd21a. 

> The -rsigopt does not change the hash algorithm for Certificate ID in the response. -rcid changes the hash algorithm for Certificate ID in the response in addition to changing response signature algorithm.

So the two options are complementary, but only if -rcid does not change the signature algorithm (i.e. only changes the Certificate ID).

**Comment by tmshort** [comment]
The default signing algorithm since 1.0.2 is SHA256. Our tester didn't realize that, and thought that the -rcid option updated the signing algorithm as well.

**Comment by tmshort** [comment]
@richsalz are you still looking for changes to be made?
@mattcaswell this has not been labeled for a release.

**Comment by richsalz** [comment]
This is important for interop testing.
Ping @openssl

**Comment by tmshort** [comment]
ping?

**Comment by tmshort** [comment]
ping @openssl ?

**Comment by tmshort** [comment]
ping @openssl ?

**Comment by tmshort** [comment]
ping?

File: apps/ocsp.c [hunk 2]
```diff
@@ -1168,20 +1183,27 @@ static void make_ocsp_response(BIO *err, OCSP_RESPONSE **resp, OCSP_REQUEST *req
             X509 *ca_cert = sk_X509_value(ca, jj);
             OCSP_CERTID *ca_id = OCSP_cert_to_id(cert_id_md, NULL, ca_cert);
 
-            if (OCSP_id_issuer_cmp(ca_id, cid) == 0)
+            if (OCSP_id_issuer_cmp(ca_id, cid) == 0) {
                 found = 1;
-
+                if (resp_md != NULL && jj == 0) {
```
Comment (mattcaswell) [inline apps/ocsp.c]: Why `jj == 0` here?

File: apps/ocsp.c [hunk 3]
```diff
@@ -1168,20 +1183,27 @@ static void make_ocsp_response(BIO *err, OCSP_RESPONSE **resp, OCSP_REQUEST *req
...
+                    OCSP_CERTID_free(cid_resp_md);
```
Comment (mattcaswell) [inline apps/ocsp.c]: Why is this free necessary? Didn't you already free this above?

Comment (tmshort) [reply apps/ocsp.c]: I wrote this so long ago... but, yeah, I don't think it's necessary.

File: apps/ocsp.c (hunk 2 above)
Comment (tmshort) [reply apps/ocsp.c]: Removed

**Comment by tmshort** [comment]
@mattcaswell I made the necessary changes

**Comment by kaduk** [comment]
[ @mattcaswell I reviewed a previous version of this patch internally and could re-review the final version if you think that's appropriate for the second review]

**Comment by mattcaswell** [comment]
@kaduk - I'd be ok with that.

File: apps/ocsp.c [hunk 4]
```diff
@@ -1225,6 +1244,7 @@ static void make_ocsp_response(BIO *err, OCSP_RESPONSE **resp, OCSP_REQUEST *req
             ASN1_TIME_free(revtm);
             ASN1_GENERALIZEDTIME_free(invtm);
         }
+        OCSP_CERTID_free(cid_resp_md);
```
Comment (kaduk) [inline apps/ocsp.c]: It would be a no-op to do so at present, but I would consider NULL-ing out ```cid``` here.

Comment (tmshort) [reply apps/ocsp.c]: True, there is also a way tweak the code slightly to avoid aliases

**Comment by tmshort** [comment]
This is good to go?

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**